Add buck builds on Azure Pipelines

Smjert opened the pull request:
First part of osquery/osquery#5620

Windows still have issues, I had to use some workarounds and even when giving all the full python paths it seems it is using something from MinGW which is installed on the VM.
I have no idea how to solve this issue.

I will also disable Travis at the end.

Fixes: https://github.com/osquery/osquery/issues/5620

theopolis commented on azure-pipelines.yml:
```diff
@@ -39,11 +39,71 @@ jobs:
+        if [ -z "$(LinuxCMakeStatus)" ] || [ -z "$(LinuxBuckStatus)" ]; then
+          exit 1
+        fi
+      displayName: "Detect macOS CMake and Buck build status"
```
Detect Linux?

Smjert replied:
oups, forgot to update that :).

theopolis commented on azure-pipelines.yml:
```diff
@@ -106,4 +224,68 @@ jobs:
+        if [ -z "$(WindowsCMakeStatus)" ] || [ -z "$(WindowsBuckStatus)" ]; then
+          exit 1
+        fi
+      displayName: "Detect macOS CMake and Buck build status"
```
Detect Windows?

theopolis commented on azure-pipelines.yml:
```diff
@@ -106,4 +224,68 @@ jobs:
+  - job: Windows
+
+    pool:
+      vmImage: 'Ubuntu-16.04'
```
Is this a placeholder?

Smjert replied:
No, it's meant to "collect" the results of the Windows jobs.
So since status badges can only be created for the entire pipeline or a single job, when you have multiple jobs to check for status, in this case two, you need a third job with the name of the job you create the status badge from, which fails if one of the other two job fails, succeeds otherwise.
I didn't find a way to avoid using a VM unfortunately...

directionless replied:
So it's a `boolean and` to create a "windows" status badge? Do we need that? (vs having both status badges)

Smjert replied:
The reason I'm doing that is simply space used on the README file to show the results.
This is also one of the reasons I unified the build and test jobs.
The table showing the statuses is already cramped with 3 statuses, with this PR we would have 6, and with the other PR for Debug builds it will become 8 initially, but when all the platforms and build systems will have then, it will be 12 status badges, which is quite a lot.

directionless replied:
Thanks for explaining!

Smjert commented:
Rebased on latest master

Smjert commented:
> Do you mind squash merging, when merging, this?

Rebased onto master, and squashed commits in one.

Though I still need help with Buck on Windows, because it's using python components from the wrong paths.
Maybe @fmanco knows something?

theopolis commented:
^ or @muffins re:windows

muffins commented:
Hrm, not entirely sure I follow the Python issues... This output:
```
Generating script.
Script contents:
C:\Python27amd64\python.exe "C:\Program Files\PackageManagement\NuGet\Packages\buck.2019.06.17.01\tools\buck.pex" build @mode/windows-x86_64/release osquery:osqueryd
========================== Starting Command Output ===========================
##[command]"C:\windows\system32\cmd.exe" /D /E:ON /V:OFF /S /C "CALL "D:\a\_temp\18d45893-fb1c-4ffe-a0b1-f6fe0a1fc7c8.cmd""
Not using buckd because watchman isn't installed.
PARSING BUCK FILES: FINISHED IN 5.9s
CREATING ACTION GRAPH: FINISHED IN 2.1s
BUILT 1/1155 JOBS 0.1s //osquery/utils:attribute
BUILT 2/1155 JOBS 0.0s //osquery/utils/macros:macros
BUILT 3/1155 JOBS 0.0s //third-party/rapidjson:rapidjson
BUILT 4/1155 JOBS 0.0s //third-party/gflags:gflags
BUILT 5/1155 JOBS 0.0s //third-party/glibc:dl
BUILT 6/1155 JOBS 0.0s //third-party/libarchive:libarchive
BUILT 7/1155 JOBS 0.2s //third-party/boost:boost
BUILT 8/1155 JOBS 0.0s //third-party/zstd:zstd
BUILT 9/1155 JOBS 0.0s //third-party/glibc:pthread
BUILT 10/1155 JOBS 0.0s //third-party/openssl:openssl
BUILT 11/1155 JOBS 0.0s //third-party/zlib:zlib
BUILT 12/1155 JOBS 0.0s //third-party/ssdeep-cpp:ssdeep-cpp
...
```
Gives me the impression that Python is being found fine for both 2 and 3. The issue that it's complaining about:
```
AssertionError: D:/a/1/s/.buckd/resources/2019.06.17.01/buck-modules-resources/python/pex.pex/.bootstrap/pkg_resources is not a subpath of D:/a/1/s/.buckd/resources/2019.06.17.01/buck-modules-resources/python/pex.pex\
```
Seems somewhat reminiscent of this [super old bug](https://github.com/pypa/setuptools/issues/129). Any chance you know more about the `C:\Python27amd64` that we're using? Or how it's managing it's pip packages?

Smjert commented:
> edit: Ahh perhaps I'm seeing now, are you meaning that the Python27 that's being pulled is just the wrong one? Is it possible to remove other installations and just role the one Python27 installation?

Well I'm not sure what's happening exactly but if you see it uses run_py from mingw.
Uninstalling mingw (where it is taking the wrong python from)  with choco (which takes 10mins or so to do so), now it tries to use python3:
```
BUILT 222/1155 JOBS 0.1s //third-party/sqlite:sqlite
Traceback (most recent call last):
  File "D:\a\1\s\.buckd\resources\2019.06.17.01\buck-modules-resources\python\pex.pex\.bootstrap\_pex\pex.py", line 328, in execute
  File "D:\a\1\s\.buckd\resources\2019.06.17.01\buck-modules-resources\python\pex.pex\.bootstrap\_pex\pex.py", line 261, in _wrap_coverage
  File "D:\a\1\s\.buckd\resources\2019.06.17.01\buck-modules-resources\python\pex.pex\.bootstrap\_pex\pex.py", line 293, in _wrap_profiling
  File "D:\a\1\s\.buckd\resources\2019.06.17.01\buck-modules-resources\python\pex.pex\.bootstrap\_pex\pex.py", line 371, in _execute
  File "D:\a\1\s\.buckd\resources\2019.06.17.01\buck-modules-resources\python\pex.pex\.bootstrap\_pex\pex.py", line 429, in execute_entry
  File "D:\a\1\s\.buckd\resources\2019.06.17.01\buck-modules-resources\python\pex.pex\.bootstrap\_pex\pex.py", line 434, in execute_module
  File "C:\hostedtoolcache\windows\Python\3.6.8\x64\lib\runpy.py", line 208, in run_module
    return _run_code(code, {}, init_globals, run_name, mod_spec)
  File "C:\hostedtoolcache\windows\Python\3.6.8\x64\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "D:\a\1\s\.buckd\resources\2019.06.17.01\buck-modules-resources\python\pex.pex\src\com\facebook\buck\features\python\make_pex.py", line 184, in <module>
  File "D:\a\1\s\.buckd\resources\2019.06.17.01\buck-modules-resources\python\pex.pex\src\com\facebook\buck\features\python\make_pex.py", line 157, in main
AttributeError: 'dict' object has no attribute 'iteritems'
```

Smjert commented:
@mossberg reminded me of a workaround we had to use in the past to have Buck working.
So removing mingw and setting python2 as first in the PATH makes the build work.

muffins commented:
@Smjert ah yes! this is how I have my setup at home, sorry I didn't lead with that! It's really dumb and I have no idea why Buck requires Python2 as well as 3, but thanks for your hard work getting that running!

directionless commented:
Travis seems to be causing issues on PRs. (At least, I don't know if it's a travis failure, or a real test failure) so if y'all are happy, I say merge!
